driver/queries.c

```diff
@@ -3052,23 +3046,199 @@ static SQLRETURN statement_len_json(esodbc_stmt_st *stmt, size_t *outlen)
 
 #define FAIL_ON_CBOR_ERR(_hnd, _cbor_err) \
 	do { \
-		if (err != CborNoError) { \
+		if (_cbor_err != CborNoError) { \
 			ERRH(_hnd, "CBOR: %s.", cbor_error_string(_cbor_err)); \
 			RET_HDIAG(_hnd, SQL_STATE_HY000, "CBOR serialization error", \
 				_cbor_err); \
 		} \
 	} while (0)
 
+static SQLRETURN statement_params_len_cbor(esodbc_stmt_st *stmt,
+	size_t *enc_len, size_t *conv_len)
+{
+	SQLSMALLINT i;
+	size_t len, l, max;
+	esodbc_rec_st *arec, *irec;
+	SQLRETURN ret;
+
+	/* Initial all-encompassing array preamble. */
+	/* ~ [] */
+	len = cbor_nn_hdr_len(stmt->apd->count);
+
+	max = 0;
+	for (i = 0; i < stmt->apd->count; i ++) {
+		assert(stmt->ipd->count == stmt->apd->count);
+		arec = &stmt->apd->recs[i];
+		irec = &stmt->ipd->recs[i];
+
+		/* ~ {} */
+		len = cbor_nn_hdr_len(stmt->apd->count);
+
+		/* ~ "type": "..." */
+		len += cbor_str_obj_len(sizeof(REQ_KEY_PARAM_TYPE) - 1);
+		len += cbor_str_obj_len(irec->es_type->type_name.cnt);
+
+		/* ~ "value": "..." */
+		len += cbor_str_obj_len(sizeof(REQ_KEY_PARAM_VAL) - 1);
+		assert(irec->es_type);
+
+		/* assume quick maxes */
+		ret = convert_param_val(arec, irec, /*params array pos*/0,
+				/*dest: calc length*/NULL, &l);
+		if (! SQL_SUCCEEDED(ret)) {
+			return ret;
+		}
+		/* keep maximum space required for storing the converted obj */
+		if (max < l) {
+			max = l;
+		}
+		/* the values are going to be sent as strings
+		 * (see serialize_param_cbor() note) */
+		len += cbor_str_obj_len(l);
+	}
+
+	*conv_len = max;
+	*enc_len = len;
+	return SQL_SUCCESS;
+}
+
+/* Note: this implementation will encode numeric SQL types as strings (as it's
+ * reusing the JSON converters). This is somewhat negating CBOR's intentions,
+ * but: (1) it's a simplified and tested implementation; (2) the overall
+ * performance impact is negligible with this driver's currently intended
+ * usage pattern (SELECTs only, fetching data volume far outweighing
+ * query's); (3) the server will convert the received value according to the
```

Comment:
nit: "querys" should not have an apostrophe

Reply:
Thanks, addressed.